File: web-g1-shelf/test_parity.mjs
// Node half of the parity test: runs the JS matcher with the same scripted
// player as test_parity.py and prints all rows as JSON on stdout.
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { MotionMatcher, loadDB } from './js/mm.js';

const HERE = dirname(fileURLToPath(import.meta.url));
const DATA = join(HERE, 'data');

const meta = JSON.parse(readFileSync(join(DATA, 'mm.json')));
const model = JSON.parse(readFileSync(join(DATA, 'model.json')));
const binBuf = readFileSync(join(DATA, 'mm.bin'));
const bin = binBuf.buffer.slice(binBuf.byteOffset, binBuf.byteOffset + binBuf.byteLength);

const A = loadDB(meta, bin);
const mm = new MotionMatcher(meta, A, model.bodies);

const N_TICKS = 400;
const TRIGGER_TICK = 30;

const rows = [];
for (let tick = 0; tick < N_TICKS; tick++) {
  if (tick === TRIGGER_TICK) mm.triggerPick();
  const vel = mm.held ? [-0.9, 0, 0] : [0, 0, 0];
  const face = mm.held ? [-1, 0, 0] : [0, 0, 0];
  const q = mm.step(vel, face);
  rows.push([...q, ...mm.vasePos, mm.held ? 1 : 0, mm.animFrame]);
}
console.log(JSON.stringify(rows));

File: web-g1-shelf/test_parity.py
"""Python <-> JS parity test: both matchers run the same scripted player
(press B at tick 30, carry the bottle away once held, press N at tick 250
to put it back) and every tick's qpos, vase pose, held flag and matched
frame are compared.

    python web-g1-shelf/test_parity.py     (needs node on PATH)
"""
import json
import os
import subprocess
import sys

import numpy as np

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(ROOT, "mm-g1-shelf"))

import config as C
C.APPROX_BIAS = 0.0            # exact NN, like the JS brute force

from data import load_library
from controller import MotionMatcher

N_TICKS = 650
TRIGGER_TICK = 30
PLACE_TICK = 250


def commands(tick, held):
    if held:
        return [-0.9, 0.0, 0.0], [-1.0, 0.0, 0.0]
    return [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]


def run_python():
    m = MotionMatcher(load_library())
    rows = []
    for tick in range(N_TICKS):
        if tick == TRIGGER_TICK:
            m.trigger_pick()
        if tick == PLACE_TICK:
            m.trigger_place()
        vel, face = commands(tick, m.held)
        q = m.step(np.array(vel), np.array(face))
        rows.append(list(q) + list(m.vase_pos)
                    + [1.0 if m.held else 0.0, float(m.animFrame)])
    return np.array(rows)


def run_js():
    out = subprocess.run(
        ["node", os.path.join(ROOT, "web-g1-shelf", "test_parity.mjs")],
        capture_output=True, text=True, check=True)
    return np.array(json.loads(out.stdout.splitlines()[-1]))


def main():
    py = run_python()
    js = run_js()
    assert py.shape == js.shape, (py.shape, js.shape)
    dq = np.abs(py[:, :36] - js[:, :36]).max()
    dv = np.abs(py[:, 36:39] - js[:, 36:39]).max()
    held_same = np.array_equal(py[:, 39], js[:, 39])
    frames_same = np.array_equal(py[:, 40], js[:, 40])
    print(f"max |qpos| diff:  {dq:.2e}")
    print(f"max |vase| diff:  {dv:.2e}")
    print(f"held identical:   {held_same}")
    print(f"frames identical: {frames_same}")
    ok = dq < 1e-4 and dv < 1e-4 and held_same and frames_same
    print("PASS" if ok else "FAIL")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())
